core/src/engine/EngineMultiRobot.cc

```diff
@@ -0,0 +1,2257 @@
+            int32_t NumFloatEntries = std::stoi(headerNumFloatEntries.substr(delimiter + 1));
+
+            // Deduce the parameters required to parse the whole binary log file
+            integerSectionSize = (NumIntEntries - 1) * sizeof(int32_t); // Remove Global.Time
+            floatSectionSize = NumFloatEntries * sizeof(float32_t);
+            headerSize = ((int32_t) myFile.tellg()) - START_LINE_TOKEN.size() - 1;
+
+            // Close the file
+            myFile.close();
+        }
+        else
+        {
+            std::cout << "Error - EngineMultiRobot::parseLogBinary - Impossible to open the log file. "\
+                         "Check that the file exists and that you have reading permissions." << std::endl;
+            return hresult_t::ERROR_BAD_INPUT;
+        }
+
+        FileDevice device(filename);
+        device.open(OpenMode::READ_ONLY);
+        std::vector<AbstractIODevice *> flows;
+        flows.push_back(&device);
+
+        TelemetryRecorder::getData(header,
+                                   timestamps,
+                                   intData,
+                                   floatData,
+                                   flows,
+                                   integerSectionSize,
+                                   floatSectionSize,
+                                   headerSize);
+
+        return hresult_t::SUCCESS;
+    }
+
+    hresult_t EngineMultiRobot::parseLogBinary(std::string              const & filename,
+                                               std::vector<std::string>       & header,
+                                               matrixN_t                      & logData)
+    {
+        std::vector<float64_t> timestamps;
+        std::vector<std::vector<int32_t> > intData;
+        std::vector<std::vector<float32_t> > floatData;
+        hresult_t returnCode = parseLogBinaryRaw(
+            filename, header, timestamps, intData, floatData);
+        if (returnCode == hresult_t::SUCCESS)
+        {
+            logDataRawToEigenMatrix(timestamps, intData, floatData, logData);
+        }
+        return returnCode;
+    }
+
+    vectorN_t EngineMultiRobot::computeAcceleration(std::shared_ptr<Robot> robot,
+                                                    vectorN_t const & q,
+                                                    vectorN_t const & v,
+                                                    vectorN_t const & u,
+                                                    forceVector_t const & fext)
+    {
+        if (robot->hasConstraint())
+        {
+            // Handle kinematic constraints.
+            matrixN_t J;
+            vectorN_t drift;
+            robot->computeConstraints(q, v, J, drift);
+
+            // TODO: handle external forces.
+            // TODO: rotor inertia.
+
+            // Call forward dynamics.
+            float64_t damping = 1e-12;
+            return pinocchio::forwardDynamics(robot->pncModel_,
```

Comment:
Currently the dynamics is limited and does not support two things:

 - external forces (both user-defined and spring contact forces): this can be added to u by the use of the right jacobian as input.
 - rotor inertia: this requires a re-implemtation of forwardDynamics like you did for aba

I'm working on both of these things at the moment.

Reply:
I guest all the Pinocchio "patches" should be moved in a dedicated helper file.